### PR TITLE
[RFC] :sparkles: introduce the concept of synced preferences

### PR DESCRIPTION
**Pull request by MatissJanis**
Introducing the concept of - synced preferences. Looking to sanity check myself before investing in building out the feature more and before porting local prefs over to use synced prefs.

**Why do we need this?** Currently we have local & global preferences. For those unaware what they are - quick recap.

local preferences - these initially come from the `metadata.json` file that every budget had. The user can later override the defaults and they are persisted in local storage (meaning they are device-specific). If a user clears local storage - we reset back to the original `metadata.json` file

global preferences - basically just local storage; preferences that are applicable across budget files (but not cross-device)

**So again: why create synced prefs?** The problem with local prefs is that whenever you export your budget - a new `metadata.json` file is produced with your current local pref state. For example: if you switch the budget back to year 2020 and export - the active month will be persisted as "2020". Whenever you open up the budget file - the active month will be "2020". Each time you clear your cache (or use a different device) - again you will get the year "2020". _That is annoying_.

**Ok, but why not just remove the prefs when exporting?** That could be done. But arguably that would be quite annoying as well. Imagine - you configured your number format to use "1 000.33" format. And perhaps changed the language to "DE" (thinking about the future). But each time you import your budget - all these configs get reset back to their original state.

---

So.. I'm looking for a sanity-check.

1. Do you think we should have "synced preferences" (prefs that sync across device)?
2. Do you think _most_ (if not all) prefs should use these new synced prefs instead of local prefs?
3. Any other thoughts?

### CODE REVIEW

File: packages/desktop-client/src/components/App.tsx [hunk 1]
```diff
@@ -33,7 +33,6 @@ import { DevelopmentTopBar } from './DevelopmentTopBar';
 import { FatalError } from './FatalError';
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/App.tsx:33]: The code in this PR is a bit messy.. My intention here is not to merge anything. But just share a very early draft of the concept of "synced prefs". Just to get early feedback before investing in developing it more.

**Comment by MatissJanis** [comment]
The concept outlined here is my attempt at solving https://github.com/actualbudget/actual/issues/2645

**Comment by youngcw** [comment]
I think that the only prefs that should sync between devices are those that affect the budget itself.  Like number format, language, budget type, etc.  These could even be saved in the db instead of having the metadata.json file.

Then everything else probably should stay per device.  Such as what month you last looked at, number of months you were viewing, sidebar width, etc.  These GUI based ones could persist between files, but don't need to either.

**Comment by matt-fidd** [comment]
> I think that the only prefs that should sync between devices are those that affect the budget itself. Like number format, language, budget type, etc. These could even be saved in the db instead of having the metadata.json file.
> 
> Then everything else probably should stay per device. Such as what month you last looked at, number of months you were viewing, sidebar width, etc. These GUI based ones could persist between files, but don't need to either.

To add to this, I'd appreciate if transaction table view settings like which accounts the running balance shows on etc... were synced too.

**Comment by MatissJanis** [comment]
That makes sense.

I've compiled a list of all our current preferences and how they are used here. Along with where I would propose we move them in the near future (i.e. synced vs local storage vs existing metadata.json). Please feel free to take a look and let me know your thoughts via comments in the file: https://docs.google.com/spreadsheets/d/1fIuHnWdJa0lCT6KhinvlnApZIwv-_yLMUcrAqiizjww/edit?usp=sharing


**Why keep using metadata.json for some of these?** Ideally I would like to move everything to either synced OR local storage, but porting some of these will be very difficult. So in stage-1 of the preference clean-up I plan on separating the majority of the prefs. In stage 2 we can talk about getting rid of `metadata.json` too. Lets focus on stage 1 for now to limit the scope.

**Comment by MatissJanis** [comment]
One more thought: we kinda already have some logic that periodically uploads the `metadata.json` file. Theoretically that is the "sync" option except it's much slower and less reliable. Alternative path we could take here is - instead of creating a new concept of synced prefs - look into patching the `upload` method so it would work. And then also move some of the preferences to local storage (so they would not be stored in metadata.json).

https://github.com/actualbudget/actual/blob/d032fce7ea701aa054796a6b54eb914276ac6e08/packages/loot-core/src/server/cloud-storage.ts#L318

**Comment by twk3** [comment]
I'm in agreement with what has already been mentioned above

>  1. Do you think we should have "synced preferences" (prefs that sync across device)?

Yep, I think that makes sense, and would be desired for most folks using multiple devices to access their budget

> 2. Do you think most (if not all) prefs should use these new synced prefs instead of local prefs?

Most, but not all, as others mentioned. I reviewed the google doc, and the ones in there left as local prefs made sense to me. Their all UI changes (other than the new version notification), and ones that might be impactful based on your device/type screen size.

**Pull request closed without merging**